Sources/SwiftParser/Statements.swift

```diff
@@ -877,11 +918,66 @@ extension Parser.Lookahead {
         // For example, could be the function call "discard()".
         return false
       }
+
+    case .then:
+      return isStartOfThenStatement(preferExpr: preferExpr)
+
     case nil:
       return false
     }
   }
 
+  /// Whether we're currently at a `then` token that should be parsed as a
+  /// `then` statement.
+  mutating func isStartOfThenStatement(preferExpr: Bool) -> Bool {
+    // If the feature is disabled, don't parse it.
+    guard self.experimentalFeatures.contains(.thenStatements) else {
+      return false
+    }
+    guard self.at(.keyword(.then)) else {
+      return false
+    }
+
+    // If we want to prefer an expr, and aren't at the start of a newline, then
+    // don't parse a ThenStmt.
+    if preferExpr && !self.atStartOfLine {
+      return false
+    }
+
+    let next = peek()
+
+    // If 'then' is followed by a binary or postfix operator, prefer to parse as
+    // an expr.
+    if BinaryOperatorLike(lexeme: next) != nil || PostfixOperatorLike(lexeme: next) != nil {
+      return false
+    }
+
+    switch next.rawTokenKind {
+    case .keyword:
+      // Treat 'is' and 'as' like the binary operator case, and parse as an
+      // expr.
+      switch Keyword(next.tokenText) {
+      case .as, .is:
+        return false
+      default:
+        break
+      }
+
+    case .leftBrace:
+      // This is a trailing closure.
+      return false
+
+    case .leftParen, .leftSquare, .period:
+      // These are handled based on whether there is trivia between the 'then'
+      // and the token. If so, it's a 'then' statement. Otherwise it should
+      // be treated as an expression, e.g `then(...)`, `then[...]`, `then.foo`.
+      return !self.currentToken.trailingTriviaText.isEmpty || !next.leadingTriviaText.isEmpty
```

Comment:
AFAICT, the following would be interpreted as a `ThenStmt` but I feel like it would be fairly common to write

```swift
then
  .filter { ... }
  .map { ... }
```